### PR TITLE
checks core table and audit tables are unique

### LINKED ISSUE #2542: Audit session upload failed from proxy server
Have a question? Please use [Our Forum](https://community.gravitational.com)

**What happened**:
Audit sessions failed to upload from proxy while running a proxy recording set up with DynamDB and S3 bucket backend for audit logs.

Proxy server is complaining that `HashKey` is missing. I believed I have supplied all the values specified in the documentation on using DynamoDB and S3 as backend.

One thing I don't get is why is session uploaded failure attribute to missing DynamoDB hashkey. Aren't sessions uploaded to S3?

**What you expected to happen**:
Session upload should succeed.

**How to reproduce it (as minimally and precisely as possible)**:

**Auth config**
```
teleport:
    nodename: auth-1
    data_dir: /var/lib/teleport
    advertise_ip: <redacted>
    auth_servers:
        - <redacted>

    storage:
        type: dynamodb
        region: ap-southeast-1
        table_name: dynamodb.table.name

        audit_events_uri: dynamodb://dynamodb.table.name
        audit_sessions_uri: s3://<redacted>

auth_service:
    enabled: True
    cluster_name: primary
    authentication:
        second_factor: u2f
        type: local
        u2f:
            app_id: <redacted>
            facets:
            - <redacted>

    listen_addr: 0.0.0.0:3025
    session_recording: proxy

ssh_service:
    enabled: True
    listen_addr: 0.0.0.0:3022
```

**Proxy config** 
```
teleport:
    nodename: proxy-1
    data_dir: /var/lib/teleport
    advertise_ip: <redacted>
    auth_servers:
        - <redacted>
    connection_limits:
        max_connections: 1000
        max_users: 250

ssh_service:
    enabled: True
    listen_addr: 0.0.0.0:3022

proxy_service:
    enabled: True
    listen_addr: 0.0.0.0:3023
    web_listen_addr: 0.0.0.0:3080
    tunnel_listen_addr: 0.0.0.0:3024
    public_addr: <redacted>
    https_key_file: <redacted>
    https_cert_file: <redacted>
```

**Environment**:
- Teleport version (use `teleport version`): `Teleport Enterprise v3.1.4git:v3.1.4-0-g9b7b5b8b go1.11.4`
- Tsh version (use `tsh version`): `Teleport v3.1.4 git:v3.1.4-0-g9b7b5b8b go1.11.4`
- OS (e.g. from /etc/os-release): for both auth and proxy server
```
NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

**Browser environment**

- Browser Version (for UI-related issues):
- Install tools:
- Others:

**Relevant Debug Logs If Applicable**
```
Feb 01 04:39:00 ch-teleport-proxy-1 teleport[1025]: WARN [AUDIT]     "Session upload failed: 
ERROR REPORT:
Original Error: *errors.errorString {\"message\":\"ValidationException: One or more parameter values were invalid: Missing the key HashKey in the item
    status code: 400, request id: 8TN75UHQ94HIEETCL58TFMF4SBVV4KQNSO5AEMVJF66Q9ASUAAJG\"}
Stack Trace:
    /gopath/src/github.com/gravitational/teleport/vendor/github.com/gravitational/trace/httplib.go:78 github.com/gravitational/teleport/vendor/github.com/gravitational/trace.ReadError
    /gopath/src/github.com/gravitational/teleport/lib/httplib/httplib.go:114 github.com/gravitational/teleport/lib/httplib.ConvertResponse
   /gopath/src/github.com/gravitational/teleport/lib/auth/clt.go:256 github.com/gravitational/teleport/lib/auth.(*Client).PostForm
    /gopath/src/github.com/gravitational/teleport/lib/auth/clt.go:1712 github.com/gravitational/teleport/lib/auth.(*Client).UploadSessionRecording
    /gopath/src/github.com/gravitational/teleport/lib/events/uploader.go:243 github.com/gravitational/teleport/lib/events.(*Uploader).uploadFile.func1
    /opt/go/src/runtime/asm_amd64.s:1334 runtime.goexit
User Message: {\"message\":\"ValidationException: One or more parameter values were invalid: Missing the key HashKey in the item
    status code: 400, request id: 8TN75UHQ94HIEETCL58TFMF4SBVV4KQNSO5AEMVJF66Q9ASUAAJG\"}" duration:107.780836ms session-id:a1b1079d-1b00-11e9-b051-02a049aa03ca events/uploader.go:254
```

### PR DESCRIPTION
**Pull request by stevenGravy**
fixes #2542

Checks firestore and dynamodbo configurations do not have the same tables (collections for firestore) in core and audit events.

### CODE REVIEW

File: lib/backend/dynamo/dynamodbbk.go [hunk 1]
```diff
@@ -97,11 +100,20 @@ type Config struct {
 // CheckAndSetDefaults is a helper returns an error if the supplied configuration
 // is not enough to connect to DynamoDB
 func (cfg *Config) CheckAndSetDefaults() error {
+
```
Comment (zmb3) [inline lib/backend/dynamo/dynamodbbk.go:103]: Why add an extra line here?

File: lib/backend/dynamo/dynamodbbk.go [hunk 2]
```diff
@@ -97,11 +100,20 @@ type Config struct {
...
 	// Table name is required.
 	if cfg.TableName == "" {
 		return trace.BadParameter("DynamoDB: table_name is not specified")
 	}
 
+	// Check the TableName is not being used within the audit uri list
+	dynamoDBURI := regexp.MustCompile(`((?i)dynamodb)://` + cfg.TableName)
```
Comment (zmb3) [inline lib/backend/dynamo/dynamodbbk.go:110]: This variable does not hold a URI, so I don't think the name is appropriate.

File: lib/backend/dynamo/dynamodbbk.go [hunk 3]
```diff
@@ -97,11 +100,20 @@ type Config struct {
...
+	for _, auditURI := range cfg.AuditEventsURIs {
+		if dynamoDBURI.MatchString(auditURI) {
+			return trace.BadParameter(("DynamoDB: Same table used for table_name and audit_events_uri. Separate DynamoDB tables are required."))
```
Comment (zmb3) [inline lib/backend/dynamo/dynamodbbk.go:113]: Double (( ))?

File: lib/backend/dynamo/dynamodbbk.go (hunk 2 above)
Comment (stevenGravy) [reply lib/backend/dynamo/dynamodbbk.go]: @zmb3 renamed here and in the firestore check

**Comment by stevenGravy** [comment]
Should make these just error logging for now?

**Review by fspmarshall — CHANGES_REQUESTED** [review]
What @rosstimothy said, this could break existing deployments.  We've encountered multiple users that have made this mistake, and while they experience issues their clusters work.  We don't want their auth servers to just refuse to start post-update.

We should emit a warning, and if possible refuse to create _new_ storage with duplicate names, but we need to take care to not break existing deployments.

**Comment by stevenGravy** [comment]
> What @rosstimothy said, this could break existing deployments.  We've encountered multiple users that have made this mistake, and while they experience issues their clusters work.  We don't want their auth servers to just refuse to start post-update.
> 
> 
> 
> We should emit a warning, and if possible refuse to create _new_ storage with duplicate names, but we need to take care to not break existing deployments.

I see, will look to put that in.

**Comment by stevenGravy** [comment]
Closing for now as it doesn't seem something we could enforce right now

**Pull request closed without merging**